### ISSUE TITLE
OpenSim::Model::finalizeFromProperties leaks memory?

### ISSUE THREAD

**Issue by adamkewley** [opensim-org/opensim-core#2944]
A program like this (current `master` branch):

```c++
#include <OpenSim.h>

int main(int argc, char** argv) {
    OpenSim::Model m{"models/RajagopalModel/Rajagopal2015.osim"};
    for (size_t i = 0; i < 10; ++i) {
        OpenSim::Model m2{m};
        m2.finalizeFromProperties();  // required for the leak
    }
    return 0;
}
```

Leaks *a lot* (>5 MB per iteration) of memory. The relevance of this is when using OpenSim in a long-running application like a GUI, where models may be copied a fair bit (e.g. to launch simulations, undo-redo storage, etc.).

The memory being leaked was allocated during the copy into `m2`. Many leaks look like this:

```text
==9501== 27,180,942 (628,320 direct, 26,552,622 indirect) bytes in 374 blocks are definitely lost in loss record 1,237 of 1,237
==9501==    at 0x4835DEF: operator new(unsigned long) (vg_replace_malloc.c:334)
==9501==    by 0x4F1448D: OpenSim::Millard2012EquilibriumMuscle::clone() const (Millard2012EquilibriumMuscle.h:202)
==9501==    by 0x4AAF9A7: operator= (ArrayPtrs.h:212)
==9501==    by 0x4AAF9A7: OpenSim::ArrayPtrs<OpenSim::Actuator>::operator=(OpenSim::ArrayPtrs<OpenSim::Actuator> const&) (ArrayPtrs.h:197)
==9501==    by 0x4AB8A74: OpenSim::Set<OpenSim::Actuator, OpenSim::Object>::Set(OpenSim::Set<OpenSim::Actuator, OpenSim::Object> const&) (Set.h:116)
==9501==    by 0x5433B4C: ForceSet (ForceSet.h:48)
==9501==    by 0x5433B4C: clone (ForceSet.h:49)
==9501==    by 0x5433B4C: cloneOrNull (ClonePtr.h:312)
==9501==    by 0x5433B4C: ClonePtr (ClonePtr.h:93)
==9501==    by 0x5433B4C: copyConstruct<const SimTK::ClonePtr<OpenSim::ForceSet>*> (Array.h:3182)
==9501==    by 0x5433B4C: Array_ (Array.h:1657)
==9501==    by 0x5433B4C: ObjectProperty (Property.h:1058)
==9501==    by 0x5433B4C: OpenSim::ObjectProperty<OpenSim::ForceSet>::clone() const (Property.h:1088)
==9501==    by 0x58DBAE9: OpenSim::PropertyTable::replaceProperties(SimTK::Array_<OpenSim::AbstractProperty*, unsigned int> const&) (PropertyTable.cpp:159)
==9501==    by 0x58DBF05: OpenSim::PropertyTable::operator=(OpenSim::PropertyTable const&) (PropertyTable.cpp:56)
==9501==    by 0x58A0322: operator= (Object.cpp:192)
==9501==    by 0x58A0322: OpenSim::Object::operator=(OpenSim::Object const&) (Object.cpp:185)
==9501==    by 0x58A08E0: OpenSim::Object::Object(OpenSim::Object const&) (Object.cpp:165)
==9501==    by 0x169EFF: OpenSim::Component::Component(OpenSim::Component const&) (Component.h:427)
==9501==    by 0x16A2BE: OpenSim::ModelComponent::ModelComponent(OpenSim::ModelComponent const&) (ModelComponent.h:82)
==9501==    by 0x16ACFF: OpenSim::Model::Model(OpenSim::Model const&) (MultibodyGraphMaker.h:152)
==9501==    by 0x21225B: main (osmv.cpp:55)
```

or this:

```text
==9501==    at 0x4835DEF: operator new(unsigned long) (vg_replace_malloc.c:334)
==9501==    by 0x4E7A65D: OpenSim::CoordinateActuator::clone() const (CoordinateActuator.h:49)
==9501==    by 0x4AAF9A7: operator= (ArrayPtrs.h:212)
==9501==    by 0x4AAF9A7: OpenSim::ArrayPtrs<OpenSim::Actuator>::operator=(OpenSim::ArrayPtrs<OpenSim::Actuator> const&) (ArrayPtrs.h:197)
==9501==    by 0x4AB8A74: OpenSim::Set<OpenSim::Actuator, OpenSim::Object>::Set(OpenSim::Set<OpenSim::Actuator, OpenSim::Object> const&) (Set.h:116)
==9501==    by 0x5433B4C: ForceSet (ForceSet.h:48)
==9501==    by 0x5433B4C: clone (ForceSet.h:49)
==9501==    by 0x5433B4C: cloneOrNull (ClonePtr.h:312)
==9501==    by 0x5433B4C: ClonePtr (ClonePtr.h:93)
==9501==    by 0x5433B4C: copyConstruct<const SimTK::ClonePtr<OpenSim::ForceSet>*> (Array.h:3182)
==9501==    by 0x5433B4C: Array_ (Array.h:1657)
==9501==    by 0x5433B4C: ObjectProperty (Property.h:1058)
==9501==    by 0x5433B4C: OpenSim::ObjectProperty<OpenSim::ForceSet>::clone() const (Property.h:1088)
==9501==    by 0x58DBAE9: OpenSim::PropertyTable::replaceProperties(SimTK::Array_<OpenSim::AbstractProperty*, unsigned int> const&) (PropertyTable.cpp:159)
==9501==    by 0x58DBF05: OpenSim::PropertyTable::operator=(OpenSim::PropertyTable const&) (PropertyTable.cpp:56)
==9501==    by 0x58A0322: operator= (Object.cpp:192)
==9501==    by 0x58A0322: OpenSim::Object::operator=(OpenSim::Object const&) (Object.cpp:185)
==9501==    by 0x58A08E0: OpenSim::Object::Object(OpenSim::Object const&) (Object.cpp:165)
==9501==    by 0x169EFF: OpenSim::Component::Component(OpenSim::Component const&) (Component.h:427)
==9501==    by 0x16A2BE: OpenSim::ModelComponent::ModelComponent(OpenSim::ModelComponent const&) (ModelComponent.h:82)
==9501==    by 0x16ACFF: OpenSim::Model::Model(OpenSim::Model const&) (MultibodyGraphMaker.h:152)
==9501==    by 0x21225B: main (osmv.cpp:55)
```

Full leak dump: https://gist.github.com/adamkewley/7058b2b39d02170597ec5dbf975fd8c5

The main thing to note is that *something* in `finalizeFromProperties` is (e.g.) constructing things without properly destructing the existing data. I think this because the `OpenSim::Model::~Model()` destructor seems to work fine in isolation (i.e. if I do not call `finalizeFromProperties()`).

Leak dump without `.finalizeFromProperties()`: https://gist.github.com/adamkewley/b8c94e4aaf2b2c3a79a02c069cd9708f

There are stupid leaks there without `finalizeFromProperties` (e.g. the XML parser), but nowhere near as large as with it.

**Comment by adamkewley** [comment]
After some (painful) bisecting, I've drilled to:

- `Component::finalizeFromProperties:270` [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Component.cpp#L273)
- `Model::extendFinalizeFromProperties:685` [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/Model/Model.cpp#L701)
- `ForceSet::updActuators:190` [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/Model/ForceSet.cpp#L216)
- `Set::adoptAndAppend` [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Set.h#L462)

```c++
void ForceSet::updateActuators() {
    // GOOD
    _actuators.setMemoryOwner(false);
    _actuators.setSize(0);
    // GOOD
    for (int i = 0; i < getSize(); ++i) {
        Actuator* act = dynamic_cast<Actuator*>(&get(i));
        if (act) _actuators.adoptAndAppend(act);
    }
    // BAD
}
```

Seems pertinent, but needs more understanding.

**Comment by adamkewley** [comment]
So the issue here is that `OpenSim::ArrayPtrs` has really unclear ownership semantics. 

From a class design PoV, it's:

- A dynamically sized array (much like vector) ([impl](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ArrayPtrs.h))
- Specifically for pointers
- Which it may, or may not own ([code](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ArrayPtrs.h#L72))
- Which copy-constructs by assignment ([code](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ArrayPtrs.h#L133))
- Where assignment has the semantics of clearing + destroying memory only if it's the owner ([here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ArrayPtrs.h#L200))
- But, **regardless of the other Array's ownership status**, will `.clone()` (effectively, heap-copy) every pointer in the right-hand side of the assignment and take ownership of the contents ([here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ArrayPtrs.h#L212))
- But callers may (and, in this case, do) call `setMemoryOwner(false)` on the assign-constructed array that *definitely* owns its contents ([here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ArrayPtrs.h#L316))

So, stage set. The leak above is this:

- Create a model
- Copy-construct another model
- C++ semantics /w auto-generated copy constructors copy-constructs `Model::ForceSet`
- This recurses through to `ForceSet::ForceSet` (`forceSet`), which recurses to `Set<Actuator>` (`_actuators`), which recurses to `PropObjArray<Actuator>` (`_propObjects`), which recurses to `ArrayPtrs<Actuator>` (`_array`)
- This short-cuts to `ArrayPtrs::operator=(const ArrayPtrs<Actuator>& aArray)`
- The array is empty (because it was only just constructed)
- `ArrayPtrs::operator=(const ArrayPtrs<Actuator>& aArray)` copies all the `Actuator`s with deep-copy semantics and takes ownership of those copies

Now we have a copied model, but one of these things is not like the other:
- The original `Model` was holding onto pointers (effectively, cross-references) to `Actuator`s. 
- The copy is holding onto independent clones of `Actuator`s. 
- When `Model::finalizeProperties` is called, it calls (via the steps outlined from bisecting) into `forceSet.updActuators`, which calls `.setMemoryOwner(false)`.
- The copy now owns independent clones of `Actuator`s which will not be destroyed when the Model is destructed :boom:

**Comment by tgeijten** [comment]
After applying #2951, the leak is gone, but the application has become very unstable. If I run an OpenSim4 optimization in SCONE, the application will crash at seemingly random moments.

In comparison, in a pre #2951 build, there are none of these crashes -- except when at some point the system runs out of memory.

Perhaps the memory is released too early and still referenced somewhere..?

**Comment by tgeijten** [comment]
I have found a workaround: putting a mutex around the copied model until _after_ the call to `initSystem()` appears to keep it from crashing. Not sure why this would be the case. This workaround does affect performance, since different simulations now have to wait on each other longer during construction.

**Comment by adamkewley** [comment]
@tgeijten in what circumstances are you sharing a model between threads? E.g. what methods are you calling under your mutex?

**Comment by adamkewley** [comment]
@tgeijten , if [this](https://github.com/opensim-org/SCONE/blob/f20f73651e09899e009a47a7892f90883a6a9413/src/sconelib/sconeopensim4/ModelOpenSim4.cpp#L116) code section is what you now need to protect with a lock, I'm surprised you didn't get crashes earlier. 

A quick scroll through that code section revealed at least 3 mutations to the OpenSim model  ([here](https://github.com/opensim-org/SCONE/blob/f20f73651e09899e009a47a7892f90883a6a9413/src/sconelib/sconeopensim4/ModelOpenSim4.cpp#L138), [here](https://github.com/opensim-org/SCONE/blob/f20f73651e09899e009a47a7892f90883a6a9413/src/sconelib/sconeopensim4/ModelOpenSim4.cpp#L149), and [here](https://github.com/opensim-org/SCONE/blob/f20f73651e09899e009a47a7892f90883a6a9413/src/sconelib/sconeopensim4/ModelOpenSim4.cpp#L178)). If that code section worked before, but without a mutex, it was almost certainly a happy accident. 

Pre-PR, you might have gotten lucky (ignoring the leak) because the leak was causing copies to hold onto unique, independent, `.clone`s of things, rather than aliasing the same memory location - no idea.

OpenSim objects are, by no stretch of the imagination, threadsafe. I personally follow the rule that even `const` methods are not threadsafe, because there are `const_cast` tricks in parts of the codebase (for "logical" constness) which, violate thread safety (e.g. it's "logically" const to update a cache, but that cache might not be threadsafe).

**Comment by tgeijten** [comment]
@adamkewley the weird thing is that it happens while I'm working on an exclusive copy of a model -- not the shared model which I'm copying (I was already using a mutex for that). Specifically, I need to hold a global mutex until _after_ I call `initSystem()` on the copy. So until that moment there's somehow still a reference to some shared data that can be messed up by other threads.

**Comment by tgeijten** [comment]
So I missed your latest comment, but just to clarify: the mutations you refer to are all from a single thread. Each thread operates on its own copy of the model, the only shared data is the original model that’s being copied.

**Comment by adamkewley** [comment]
@tgeijten ah, I see :+1: . I wasn't familiar with the file resource cache returning a *copy*, rather than the cached object (reference).

Since this is an independent copy-constructed model, yeah, this should be fine. The fact that it isn't fine is worrying. Figuring out why might be a slog, because the PR that breaks it (#2951 ) is fairly small but mostly innocent enough (granted, the devil is in the details ;)).

The `ForceSet` has a fairly large amount of custom container classes + inherited data. Although it only *directly* has two members (`_actuators`, `_muscles`), it's more like: `ForceSet + ModelComponentSet<Force> + Set<Force, ModelComponent> + ModelComponent + Component + Object`, where data members from anything in that hierarchy may not strictly copy-construct into an entirely independent copy. All it would take is (e.g.) a `std::vector<T*>` to produce cross-copy aliasing.

Pre #2951, this may have been ignore-able because the `ForceSet` was aggro-`.clone()`ing  things that it should have been aliasing (the `.clone()` may de-alias something that's otherwise not copy-constructed correctly). The fact that the aliasing is "fixed" by calling `.initSystem` implies that there is a step in `initSystem` that (e.g.) wipes an alias and replaces it with something valid. This is probable: the `ForceSet` leak was entirely because `finalizeFromProperties` was being used to update aliases in `ForceSet`.

**Comment by tgeijten** [comment]
Great 😄 I suppose `xo::resource_cache` should get a `clone()` method instead `operator()()`.

I realize #2951 isn't the cause of the problem; it only revealed an underlying issue that was present all along, but was masked by the memory leak. Shall I create a new issue for this?

**Comment by adamkewley** [comment]
Either or. It needs to be fixed, but the new issue seems to be related to aliasing between copies, rather than a leak.

I could PR a reverse-patch to master if it's deemed that aliasing is worse than leaking, but the leaks are fairly annoying though for anything long-running

**Comment by tgeijten** [comment]
The leaks are far worse than the aliasing, especially now that I have a workaround.

**Comment by tgeijten** [comment]
I've created a new issue for the crash. I'm not sure if this issue can be closed, since the current hotfix #2951 might not be sufficient?

**Comment by adamkewley** [comment]
I have recently fixed almost all of the usual leaks, apart from one in the XML parser, which will require more work
